### PR TITLE
fix(bqstream): event ordering

### PR DESCRIPTION
**Pull request by shrouti1507**
## Description of the change

> resolves INT-389

Fixing event ordering issue for BQ stream. 
Dividing events to different user journeys and keeping the order of the events intact.

## Type of change
- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: test/integrations/destinations/bqstream/router/data.ts [hunk 1]
```diff
@@ -153,6 +154,7 @@ export const data = [
               },
               metadata: {
                 jobId: 2,
+                userId: 'user12345',
```
Comment (mihir-4116) [inline test/integrations/destinations/bqstream/router/data.ts:157]: add tests with multiple users

**Review by mihir-4116 — APPROVED** [review]

File: test/integrations/destinations/bqstream/router/data.ts (hunk 1 above)
Comment (shrouti1507) [reply test/integrations/destinations/bqstream/router/data.ts:157]: added

**Review by mihir-4116 — APPROVED** [review]

File: src/v0/destinations/bqstream/transform.js [hunk 2]
```diff
@@ -102,38 +103,52 @@ const processRouterDest = (inputs) => {
...
+              eventsChunk.push({
+                message: res,
+                metadata: event.metadata,
+                destination: event.destination,
+              });
+            });
+          }
+        } catch (error) {
+          const errRespEvent = handleRtTfSingleEventError(event, error, DESTINATION);
+          // divide the successful payloads till now into batches
```
Comment (koladilip) [inline src/v0/destinations/bqstream/transform.js:131]: why do we need to batch this way? is it not possible process events all and then handling batching?

Comment (shrouti1507) [reply src/v0/destinations/bqstream/transform.js:131]: Yes! As you said, we were handling it just like this previously. 

example : 
u1 is sending -->  [ track1, track2, identify1, track3, track 4, identify 4, identify 5, track 5, track 6 ], now, the transformer was sending [[ track1, track2, track3, track 4, track 5  ], [identify1],  [identify 4], [ identify 5 ]]. 

In the case of the bq stream, the identify event is not supported.  But, to maintain the event order and user journey, we need to break the batches anyway. For example,

u1 is sending -->  [ track1, track2, identify1, track3, track 4, identify 4, identify 5, track 5 ], now, to maintain the above scenarios, the transformer is sending [[ track1, track2 ], [identify1], [track3, track 4], [identify 4], [ identify 5 ], [ track 6]]. 

With little change, we could do it this way. Do you recommend keeping the catch part only for error-catching and writing an event order handling util function to handle this?

Comment (koladilip) [reply src/v0/destinations/bqstream/transform.js:131]: code is hard to follow now. what ever may be requirement, code should be easy to understand

**Review by koladilip — APPROVED** [review]

File: src/v0/destinations/bqstream/transform.js (hunk 2 above)
Comment (shrouti1507) [reply src/v0/destinations/bqstream/transform.js]: addressed

**Review by koladilip — APPROVED** [review]

File: test/integrations/destinations/bqstream/router/data.ts [hunk 3]
```diff
@@ -225,6 +391,76 @@ export const data = [
...
+                errorCategory: 'dataValidation',
+                errorType: 'instrumentation',
+                feature: 'router',
+                implementation: 'native',
+                module: 'destination',
+              },
+              statusCode: 400,
+            },
+            {
+              batched: true,
```
Comment (chandumlg) [inline test/integrations/destinations/bqstream/router/data.ts:428]: In this test case, only one event is batched. Can we have a some better test cases? 

For ex, Have multiple test events of different users as input and output should be batched events satisfying user level event order constraints.

File: src/v0/destinations/bqstream/utils.test.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+describe('HandleEventOrdering', () => {
+
+    // Tests that the function returns an array of transformed events when there are no error events
+    it('should return an array of transformed events when there are no error events', () => {
+      const eachUserSuccessEventslist = [{message: { type: "track"}, metadata: {jobId: 1}}, {message: { type: "track"}, metadata: {jobId: 3}}, {message: { type: "track"}, metadata: {jobId: 5}}];
+      const eachUserErrorEventsList = [];
+      const expected =  [[{"message":{"type":"track"},"metadata":[{"jobId":1}]},{"message":{"type":"track"},"metadata":[{"jobId":3}]},{"message":{"type":"track"},"metadata":[{"jobId":5}]}]];
+      const result = HandleEventOrdering(eachUserSuccessEventslist, eachUserErrorEventsList);
+      console.log(JSON.stringify(result))
```
Comment (chandumlg) [inline src/v0/destinations/bqstream/utils.test.js:11]: Remove log

Comment (shrouti1507) [reply src/v0/destinations/bqstream/utils.test.js]: done

File: test/integrations/destinations/bqstream/router/data.ts (hunk 3 above)
Comment (shrouti1507) [reply test/integrations/destinations/bqstream/router/data.ts:742]: added

File: test/integrations/destinations/bqstream/router/data.ts [hunk 5]
```diff
@@ -225,6 +705,157 @@ export const data = [
...
+                ID: '1WXjIHpu7ETXgjfiGPW3kCUgZFR',
+                Name: 'bqstream test',
+              },
+              metadata: [
+                {
+                  jobId: 5,
+                  userId: 'user123',
+                },
+                {
+                  jobId: 6,
```
Comment (chandumlg) [inline test/integrations/destinations/bqstream/router/data.ts:783]: jobId: 6 is repeated. Can you make sure jobIds are provided correctly.

Also, can we trim down the test case with not needed fields. Very difficult to read.

Comment (shrouti1507) [reply test/integrations/destinations/bqstream/router/data.ts:424]: done

**Comment by shrouti1507** [comment]
The PR working tree got disrupted after https://github.com/rudderlabs/rudder-transformer/pull/2558/commits/2c77215e963afc2a153672cdfec1a0a6811f9a54. We have opened a new PR mentioned above. So closing this PR.
